src/main/kotlin/com/exactpro/th2/rptdataprovider/handlers/messages/MessageExtractor.kt

```diff
@@ -138,139 +99,180 @@ class MessageExtractor<B, G, RM, PM>(
-                                    request.endTimestamp?.let { builder.timestampFrom().isGreaterThan(it) }
-                                }
-                            }
-                        }.build()
+                val sessionGroup = context.cradleService.getSessionGroup(
+                    commonStreamName.bookId,
+                    commonStreamName.name,
+                    request.startTimestamp,
+                    request.endTimestamp,
+                )
+                val cradleMessageIterable = context.cradleService.getGroupedMessages(
+                    this,
+                    GroupedMessageFilterBuilder().apply {
+                        bookId(commonStreamName.bookId)
+                        groupName(sessionGroup)
+                        order(order)
+
+                        if (order == REVERSE) { // default: DIRECT
+                            request.startTimestamp?.let { timestampTo().isLessThanOrEqualTo(it) }
+                            request.endTimestamp?.let { timestampFrom().isGreaterThan(it) }
+                        } else {
+                            request.startTimestamp?.let { timestampFrom().isGreaterThanOrEqualTo(it) }
+                            request.endTimestamp?.let { timestampTo().isLessThan(it) }
+                        }
+                    }.build()
                 )
 
-                logger.debug { "cradle iterator has been built for stream $streamName" }
+                LOGGER.debug { "cradle iterator has been built for session group: $sessionGroup, alias: $commonStreamName" }
 
-                var isLastMessageTrimmed = false
+                val start = request.startTimestamp
+                val end = request.endTimestamp
 
-                for (batch in cradleMessageIterable) {
-                    if (externalScope.isActive && !isLastMessageTrimmed) {
+                var firstNotFound = resumeFromId != null || start != null
+                var lastNotFound = true
 
+                for (batch: StoredGroupedMessageBatch in cradleMessageIterable) {
+                    if (externalScope.isActive && lastNotFound) {
                         val timeStart = System.currentTimeMillis()
 
-                        pipelineStatus.fetchedStart(streamName.toString())
-                        logger.trace { "batch ${batch.id.sequence} of stream $streamName with ${batch.messageCount} messages (${batch.batchSize} bytes) has been extracted" }
+                        pipelineStatus.fetchedStart(commonStreamName.toString())
+                        LOGGER.trace { "batch ${batch.firstTimestamp} of group $sessionGroup for stream $commonStreamName with ${batch.messageCount} messages (${batch.batchSize} bytes) has been extracted" }
+
+                        // TODO: this is not optimal solution (BEGIN)
+                        val isBatchOrdered = batch.messages.asSequence()

```

Comment:
maybe sort messages every time?
Current solution
* good: we iterate and filter messages two times, compare message timestamps ones, create list of timestamps
* bad: we iterate and filter messages two times, compare message timestamps ones, create list of timestamps, create list new messages.
Sort every time solution
* iterate, filter messages and collect to separate list only once